v0.8.27 has load failures for 3 bundled extensions (background-task-tool, lsp, teams-tool)

I've tried to install Tallow a couple of times (a week or two apart), but I got load errors from bundled extensions on both occasions. The first time, I gave trying to use Tallow. This time, I'm submitting a bug report, since from my experience, dependency issues of this nature often turn out to be cases of packages being available on the maintainer's machine but not available on a fresh machine. A common issue.

On a fresh global install of `@dungle-scrubs/tallow@0.8.27` on macOS, three bundled extensions fail to load from the published package:

 - `background-task-tool`
 - `lsp`
 - `teams-tool`

This is not caused by user extensions (`~/.tallow/extensions` is empty here and the project directory itself is empty, so there are no project-level extensions).

### 1) `background-task-tool` imports from unpublished `src/`

`extensions/_shared/pid-registry.ts:26`:

```ts
  from "../../src/pid-schema.js"
```

But `package.json` only publishes dist, extensions, runtime, schemas, skills, templates, themes, so `src/pid-schema.js` is missing in the installed package. `dist/pid-schema.js` does exist.

### 2) `teams-tool` imports from unpublished `src/`

`extensions/teams-tool/sessions/spawn.ts:18`:

```ts
  from "../../../src/model-metadata-overrides.js"
```

Same issue: `src/` is not published, but `dist/model-metadata-overrides.js` exists.

### 3) lsp is missing a runtime dependency

`extensions/lsp/index.ts:41` imports:

 ```ts
   "vscode-languageserver-protocol"
 ```

...but that package is currently in `devDependencies`, not `dependencies`, so it isn’t present at runtime in the global install.

# Reproduction

I reproduced this by calling `createTallowSession()` directly from the installed package and inspecting `result.extensions.errors`.

### Errors reported:

 - `Cannot find module '../../src/pid-schema.js'`
 - `Cannot find module 'vscode-languageserver-protocol'`
 - `Cannot find module '../../../src/model-metadata-overrides.js'`

# Likely fix

 - switch those two src/... imports to published runtime paths (or move the shared runtime code somewhere package-safe)
 - move `vscode-languageserver-protocol` to `dependencies` in `package.json`

There may also be a smaller packaging issue where crash/help text points to `docs/providers.md`, but `docs/` doesn’t seem to be published either.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- **packaging:** fix 3 bundled extensions failing on fresh install — `background-task-tool`,
+  `lsp`, and `teams-tool` imported from unpublished `src/` paths; moved to `runtime/` wrappers
+- **packaging:** move `vscode-languageserver-protocol` from devDependencies to dependencies
+  so the `lsp` extension loads on fresh installs
+- **deps:** bump `@mariozechner/pi-coding-agent`, `pi-agent-core`, and `pi-ai` from 0.62.0
+  to 0.64.0 — adapt to `ModelRegistry.create()` factory and `getApiKeyForProvider()` rename
+- **edit-tool:** propagate upstream `prepareArguments` normalization so old-format
+  `{oldText, newText}` params still work after pi 0.64.0 schema change
+
 ## [0.8.27](https://github.com/dungle-scrubs/tallow/compare/tallow-v0.8.26...tallow-v0.8.27) (2026-03-21)
 
 

--- a/extensions/_shared/pid-registry.ts
+++ b/extensions/_shared/pid-registry.ts
@@ -13,17 +13,17 @@ import { spawnSync } from "node:child_process";
 import { existsSync, mkdirSync, readFileSync, unlinkSync } from "node:fs";
 import { dirname, join } from "node:path";
 import {
-	createRuntimePathProvider,
-	type RuntimePathProvider,
-} from "../../runtime/runtime-path-provider.js";
-import {
 	isPidEntry,
 	isSessionOwner,
 	type PidEntry,
 	type SessionOwner,
 	type SessionPidFile,
 	toOwnerKey,
-} from "../../src/pid-schema.js";
+} from "../../runtime/pid-schema.js";
+import {
+	createRuntimePathProvider,
+	type RuntimePathProvider,
+} from "../../runtime/runtime-path-provider.js";
 import { atomicWriteFileSync } from "./atomic-write.js";
 import { acquireFileLock } from "./file-lock.js";
 

--- a/extensions/edit-tool-enhanced/index.ts
+++ b/extensions/edit-tool-enhanced/index.ts
@@ -93,6 +93,7 @@ export default function editLive(pi: ExtensionAPI): void {
 		label: baseEditTool.label,
 		description: baseEditTool.description,
 		parameters: baseEditTool.parameters,
+		prepareArguments: baseEditTool.prepareArguments,
 
 		renderCall(args, theme) {
 			const path = args.path ?? "file";
@@ -109,7 +110,8 @@ export default function editLive(pi: ExtensionAPI): void {
 			const filePath = params.path ?? "file";
 			const effectiveCwd = ctx?.cwd ?? process.cwd();
 			const scopedEditTool = createEditTool(effectiveCwd);
-			const result = await scopedEditTool.execute(toolCallId, params, signal, onUpdate);
+			const prepared = scopedEditTool.prepareArguments?.(params) ?? params;
+			const result = await scopedEditTool.execute(toolCallId, prepared, signal, onUpdate);
 			const details = result.details as EditToolDetails | undefined;
 			const diff = details?.diff ?? "";
 			const absoluteFilename = path.isAbsolute(filePath)

--- a/extensions/prompt-suggestions/__tests__/autocomplete.test.ts
+++ b/extensions/prompt-suggestions/__tests__/autocomplete.test.ts
@@ -44,8 +44,9 @@ function createMockRegistry(
 		find(provider: string, modelId: string) {
 			return models.find((m) => m.provider === provider && m.id === modelId);
 		},
-		async getApiKey(model: Model<Api>) {
-			return keys.get(`${model.provider}/${model.id}`);
+		async getApiKeyForProvider(provider: string) {
+			const entry = [...keys.entries()].find(([k]) => k.startsWith(`${provider}/`));
+			return entry?.[1];
 		},
 		getAvailable() {
 			return models.filter((m) => keys.has(`${m.provider}/${m.id}`));
@@ -191,7 +192,7 @@ describe("resolveAutocompleteModel", () => {
 				findCalls.push(`${provider}/${modelId}`);
 				return undefined;
 			},
-			async getApiKey() {
+			async getApiKeyForProvider() {
 				return undefined;
 			},
 			getAvailable() {

--- a/extensions/prompt-suggestions/autocomplete.ts
+++ b/extensions/prompt-suggestions/autocomplete.ts
@@ -14,8 +14,8 @@ import type { Api, Model } from "@mariozechner/pi-ai";
 export interface ModelRegistryLike {
 	/** Find a model by provider and ID. */
 	find(provider: string, modelId: string): Model<Api> | undefined;
-	/** Get API key for a model. */
-	getApiKey(model: Model<Api>): Promise<string | undefined>;
+	/** Get API key for a provider. */
+	getApiKeyForProvider(provider: string): Promise<string | undefined>;
 	/** Get all models that have auth configured. */
 	getAvailable(): Model<Api>[];
 	/** Register a provider dynamically. */
@@ -102,7 +102,7 @@ export async function tryResolveModel(
 	const model = registry.find(provider, modelId);
 	if (!model) return null;
 
-	const apiKey = await registry.getApiKey(model);
+	const apiKey = await registry.getApiKeyForProvider(model.provider);
 	if (!apiKey) return null;
 
 	return { model, apiKey };
@@ -135,7 +135,7 @@ export async function resolveAutocompleteModel(
 	const available = registry.getAvailable();
 	const sorted = [...available].sort((a, b) => (a.cost?.input ?? 0) - (b.cost?.input ?? 0));
 	for (const model of sorted) {
-		const apiKey = await registry.getApiKey(model);
+		const apiKey = await registry.getApiKeyForProvider(model.provider);
 		if (apiKey) return { model, apiKey };
 	}
 

--- a/extensions/session-memory/index.ts
+++ b/extensions/session-memory/index.ts
@@ -381,7 +381,7 @@ export default function (pi: ExtensionAPI) {
 					};
 				}
 
-				const apiKey = await ctx.modelRegistry.getApiKey(model);
+				const apiKey = await ctx.modelRegistry.getApiKeyForProvider(model.provider);
 				if (!apiKey) {
 					ctx.ui.setWorkingMessage();
 					const sc = new Set(results.map((r) => r.sessionId)).size;

--- a/extensions/session-namer/index.ts
+++ b/extensions/session-namer/index.ts
@@ -188,7 +188,7 @@ export default function (pi: ExtensionAPI): void {
 				const model = findNamingModel(ctx);
 				if (!model) return;
 
-				const apiKey = await ctx.modelRegistry.getApiKey(model);
+				const apiKey = await ctx.modelRegistry.getApiKeyForProvider(model.provider);
 				if (!apiKey) return;
 
 				const prompt = buildNamingPrompt(userText, assistantText);

--- a/extensions/teams-tool/sessions/spawn.ts
+++ b/extensions/teams-tool/sessions/spawn.ts
@@ -15,7 +15,7 @@ import {
 	SessionManager,
 	SettingsManager,
 } from "@mariozechner/pi-coding-agent";
-import { applyKnownModelMetadataOverrides } from "../../../src/model-metadata-overrides.js";
+import { applyKnownModelMetadataOverrides } from "../../../runtime/model-metadata-overrides.js";
 import { getTallowHomeDir, getTallowPath } from "../../_shared/tallow-paths.js";
 import {
 	type AgentConfig,
@@ -255,7 +255,7 @@ export async function spawnTeammateSession(
 	// Use the user's tallow auth and model config so teammates inherit
 	// API keys and custom model definitions from the main session.
 	const authStorage = AuthStorage.create(getTallowPath("auth.json"));
-	const modelRegistry = new ModelRegistry(authStorage, getTallowPath("models.json"));
+	const modelRegistry = ModelRegistry.create(authStorage, getTallowPath("models.json"));
 	applyKnownModelMetadataOverrides(modelRegistry);
 	const model = modelRegistry.find(resolvedModel.provider, resolvedModel.id);
 	if (!model) {

--- a/extensions/welcome-screen/index.ts
+++ b/extensions/welcome-screen/index.ts
@@ -11,7 +11,7 @@
 import type { ExtensionAPI } from "@mariozechner/pi-coding-agent";
 import type { TUI } from "@mariozechner/pi-tui";
 import { visibleWidth } from "@mariozechner/pi-tui";
-import { TALLOW_VERSION } from "../../src/config.js";
+import { TALLOW_VERSION } from "../../runtime/config.js";
 
 /** Timeout for npm registry fetch (ms). */
 const FETCH_TIMEOUT = 4_000;

--- a/package.json
+++ b/package.json
@@ -75,24 +75,24 @@
 	"dependencies": {
 		"@clack/prompts": "^1.1.0",
 		"@dungle-scrubs/synapse": "0.1.8",
-		"@mariozechner/pi-coding-agent": "^0.62.0",
+		"@mariozechner/pi-coding-agent": "^0.64.0",
 		"@mariozechner/pi-tui": "workspace:*",
 		"@opentelemetry/api": "^1.9.0",
 		"@sinclair/typebox": "0.34.48",
 		"ai": "^6.0.116",
 		"commander": "^14.0.3",
 		"unpdf": "^1.4.0",
-		"vscode-jsonrpc": "8.2.1"
+		"vscode-jsonrpc": "8.2.1",
+		"vscode-languageserver-protocol": "3.17.5"
 	},
 	"devDependencies": {
 		"@biomejs/biome": "2.4.2",
-		"@mariozechner/pi-agent-core": "^0.62.0",
-		"@mariozechner/pi-ai": "^0.62.0",
+		"@mariozechner/pi-agent-core": "^0.64.0",
+		"@mariozechner/pi-ai": "^0.64.0",
 		"@types/node": "25.2.3",
 		"husky": "^9.1.7",
 		"lint-staged": "^16.4.0",
-		"typescript": "^5.9.3",
-		"vscode-languageserver-protocol": "3.17.5"
+		"typescript": "^5.9.3"
 	},
 	"engines": {
 		"node": ">=22"

--- a/runtime/config.ts
+++ b/runtime/config.ts
@@ -1,0 +1,7 @@
+import { resolveRuntimeModuleUrl } from "./resolve-module.js";
+
+const mod = (await import(
+	resolveRuntimeModuleUrl("config.js")
+)) as typeof import("../src/config.js");
+
+export const TALLOW_VERSION = mod.TALLOW_VERSION;

--- a/runtime/model-metadata-overrides.ts
+++ b/runtime/model-metadata-overrides.ts
@@ -1,0 +1,7 @@
+import { resolveRuntimeModuleUrl } from "./resolve-module.js";
+
+const mod = (await import(
+	resolveRuntimeModuleUrl("model-metadata-overrides.js")
+)) as typeof import("../src/model-metadata-overrides.js");
+
+export const applyKnownModelMetadataOverrides = mod.applyKnownModelMetadataOverrides;

--- a/skills/tallow-expert/SKILL.md
+++ b/skills/tallow-expert/SKILL.md
@@ -59,8 +59,8 @@ Extensions export a default function receiving `ExtensionAPI` (conventionally na
 
 #### Registration
 
-- `registerTool(tool: ToolDefinition<TParams, TDetails>)` — Register a tool that the LLM can call.
-- `registerCommand(name: string, options: Omit<RegisteredCommand, "name">)` — Register a custom command.
+- `registerTool(tool: ToolDefinition<TParams, TDetails, TState>)` — Register a tool that the LLM can call.
+- `registerCommand(name: string, options: Omit<RegisteredCommand, "name" | "sourceInfo">)` — Register a custom command.
 - `registerFlag(name: string, options: object)` — Register a CLI flag.
 - `registerMessageRenderer(customType: string, renderer: MessageRenderer<T>)` — Register a custom renderer for CustomMessageEntry.
 - `registerProvider(name: string, config: ProviderConfig)` — Register or override a model provider.
@@ -81,7 +81,7 @@ Extensions export a default function receiving `ExtensionAPI` (conventionally na
 - `getFlag(name: string)` — Get the value of a registered CLI flag.
 - `exec(command: string, args: string[], options?: ExecOptions)` — Execute a shell command.
 - `getActiveTools()` — Get the list of currently active tool names.
-- `getAllTools()` — Get all configured tools with name and description.
+- `getAllTools()` — Get all configured tools with parameter schema and source metadata.
 - `setActiveTools(toolNames: string[])` — Set the active tools by name.
 - `getCommands()` — Get available slash commands in the current session.
 - `setModel(model: Model<any>)` — Set the current model.
@@ -154,6 +154,7 @@ Extensions export a default function receiving `ExtensionAPI` (conventionally na
 - `modelRegistry` — Model registry for API key resolution
 - `model` — Current model (may be undefined)
 - `isIdle()` — Whether the agent is idle (not streaming)
+- `signal` — The current abort signal, or undefined when the agent is not streaming.
 - `abort()` — Abort the current agent operation
 - `hasPendingMessages()` — Whether there are queued messages waiting
 - `shutdown()` — Gracefully shutdown pi and exit.
@@ -178,6 +179,7 @@ Extensions export a default function receiving `ExtensionAPI` (conventionally na
 - `onTerminalInput(handler: TerminalInputHandler)` — Listen to raw terminal input (interactive mode only).
 - `setStatus(key: string, text: string)` — Set status text in the footer/status bar.
 - `setWorkingMessage(message?: string)` — Set the working/loading message shown during streaming.
+- `setHiddenThinkingLabel(label?: string)` — Set the label shown for hidden thinking blocks.
 - `setWidget(key: string, content: string[], options?: ExtensionWidgetOptions)` — Set a widget to display above or below the editor.
 - `setTitle(title: string)` — Set the terminal window/tab title.
 - `pasteToEditor(text: string)` — Paste text into the editor, triggering paste handling (collapse for large content).

--- a/src/sdk.ts
+++ b/src/sdk.ts
@@ -1553,7 +1553,7 @@ export async function createTallowSession(
 			`\x1b[33m🔐 Migrated ${migration.migratedProviders.length} auth credential(s) to secure references: ${migration.migratedProviders.join(", ")}\x1b[0m`
 		);
 	}
-	const modelRegistry = new ModelRegistry(authStorage, join(tallowHome, "models.json"));
+	const modelRegistry = ModelRegistry.create(authStorage, join(tallowHome, "models.json"));
 	applyKnownModelMetadataOverrides(modelRegistry);
 
 	// ── Runtime API key (not persisted) ──────────────────────────────────────

